File: BOJ/31_트리에서의동적계획법/15681_트리와쿼리.py
import sys
from collections import deque
sys.setrecursionlimit(int(1e9))
read = sys.stdin.readline

def makeTree(curNode, par_) :
    for node in V[curNode] :
        if node != par_ :
            child[curNode].append(node)
            # par[node] = curNode
            makeTree(node, curNode)

def cntSubtree(curNode) :
    size[curNode] = 1
    for node in child[curNode] :
        cntSubtree(node)
        size[curNode] += size[node]

N,R,Q = map(int,read().split())
V = [[] for _ in range(N+1)]
child = [[] for _ in range(N+1)]
size = [0 for _ in range(N+1)]
for _ in range(N-1) :
    x,y = map(int,read().split())
    V[x].append(y)
    V[y].append(x)

makeTree(R, -1)
cntSubtree(R)

for _ in range(Q) :
    q = int(read())
    print(size[q])

File: BOJ/31_트리에서의동적계획법/15681_트리와쿼리.cpp
#include <stdio.h>
#include <vector>
#include <string.h>

#define MAXN 100000
using namespace std;

int N,R,Q,x,y,cnt[MAXN+1];
vector<int> adjs[MAXN+1];
vector<int> child[MAXN+1];

void makeTree(int curNode, int par) {
    for(int i=0; i<adjs[curNode].size(); i++) {
        int node = adjs[curNode][i];
        if(node != par) {
            child[curNode].push_back(node);
            makeTree(node,curNode);
        }
    }
}

void cntSubtree(int curNode) {
    cnt[curNode] = 1;
    for(int i=0; i<child[curNode].size(); i++) {
        cntSubtree(child[curNode][i]);
        cnt[curNode] += cnt[child[curNode][i]];
    }
}


int main() {

    scanf("%d %d %d",&N,&R,&Q);
    for(int i=0; i<N-1; i++) {
        scanf("%d %d",&x,&y);
        adjs[x].push_back(y);
        adjs[y].push_back(x);
    }
    memset(cnt,0,sizeof(cnt));
    makeTree(R,-1);
    cntSubtree(R);

    for(int i=0; i<Q; i++) {
        scanf("%d",&x);
        printf("%d\n",cnt[x]);
    }

    return 0;
}